### ISSUE TITLE
dword ptr gs:[eax] decode error

### ISSUE THREAD

**Issue by pzread** [Cr4sh/openreil#10]
Does the openreil support segment register now ?

The binary code `0x65,0x8B,0x00` is `mov eax, dword ptr gs:[eax]`, but the openreil give me:
```
Exception occurs: panic(): Unhandled register type (I64)
```

**Comment by Cr4sh** [comment]
Unfortunately it's a known problem of libasmir that used in OpenREIL -- it supports only FS segment register for i386 code because normally it's enough to work with the regular code that was generated by HLL compiler.
In some future I planning to fix this issue.

**Comment by pzread** [comment]
`0x64,0x8B,0x00` = `mov eax,DWORD PTR fs:[eax]` also gets same error.

BTW, I try to fix this problem by add special case to `type == Ity_I64`. It gives me the following result.
```
00000000.00     STR          R_FS:16,                 ,          V_00:16  
00000000.01      OR          V_00:16,             0:16,          V_01:32  
00000000.02     STR         R_LDT:64,                 ,          V_02:64  
00000000.03     STR         R_GDT:64,                 ,          V_03:64  
00000000.04     STR         R_EAX:32,                 ,          V_04:32  
00000000.05     ADD     R_FS_BASE:32,          V_04:32,          V_05:32  
00000000.06      OR          V_05:32,             0:32,          V_06:64  
00000000.07     SHR          V_06:64,            20:64,          V_08:64  
00000000.08     AND          V_08:64,      ffffffff:64,          V_07:32  
00000000.09      EQ          V_07:32,             0:32,           V_10:1  
00000000.0a     NOT           V_10:1,                 ,           V_09:1  
00000000.0b     AND          V_06:64,      ffffffff:64,          V_11:32  
00000000.0c     LDM          V_11:32,                 ,          V_12:32  
00000000.0d     STR          V_12:32,                 ,         R_EAX:32 
```
ugly but usable

**Comment by Cr4sh** [comment]
Hmm, it's definitely requires some improvements/optimizations, thanks for pointing.

**Comment by Cr4sh** [comment]
Check latest version from repository, I wrote [some fixes for libasmir](https://github.com/Cr4sh/openreil/commit/c1bbe08c9170c0c323d6a4ad164efefc354f38a8#diff-72f4bf5d97923922aee89ef6848ba417R143) to make FS seg access code less ugly.
Some examples of how it looks now:
```
;
; asm: mov edi, dword ptr fs:[0x30]
; data (7): 64 8b 3d 30 00 00 00
;
00000000.00     STR          R_FS:16,                 ,          V_00:16
00000000.01      OR          V_00:16,             0:16,          V_01:32
00000000.02     ADD     R_FS_BASE:32,            30:32,          V_02:32
00000000.03     STR          V_02:64,                 ,          V_03:32
00000000.04     LDM          V_03:32,                 ,          V_04:32
00000000.05     STR          V_04:32,                 ,         R_EDI:32
;
; asm: mov esi, dword ptr fs:[ecx]
; data (3): 64 8b 31
;
00000007.00     STR          R_FS:16,                 ,          V_00:16
00000007.01      OR          V_00:16,             0:16,          V_01:32
00000007.02     STR         R_ECX:32,                 ,          V_02:32
00000007.03     ADD     R_FS_BASE:32,          V_02:32,          V_03:32
00000007.04     STR          V_03:64,                 ,          V_04:32
00000007.05     LDM          V_04:32,                 ,          V_05:32
00000007.06     STR          V_05:32,                 ,         R_ESI:32
;
; asm: mov dword ptr fs:[0x30], edx
; data (7): 64 89 15 30 00 00 00
;
0000000a.00     STR          R_FS:16,                 ,          V_00:16
0000000a.01      OR          V_00:16,             0:16,          V_01:32
0000000a.02     ADD     R_FS_BASE:32,            30:32,          V_02:32
0000000a.03     STR          V_02:64,                 ,          V_03:32
0000000a.04     STR         R_EDX:32,                 ,          V_04:32
0000000a.05     STM          V_04:32,                 ,          V_03:32
;
; asm: mov dword ptr fs:[ecx], edx
; data (3): 64 89 11
;
00000011.00     STR          R_FS:16,                 ,          V_00:16
00000011.01      OR          V_00:16,             0:16,          V_01:32
00000011.02     STR         R_ECX:32,                 ,          V_02:32
00000011.03     ADD     R_FS_BASE:32,          V_02:32,          V_03:32
00000011.04     STR          V_03:64,                 ,          V_04:32
00000011.05     STR         R_EDX:32,                 ,          V_05:32
00000011.06     STM          V_05:32,                 ,          V_04:32
; --------------------------------
; R_EDI = *(R_FS_BASE + 0x30)
; R_ESI = *(R_FS_BASE + R_ECX)
; *(R_FS_BASE + 0x30) = R_EDX
; *(R_FS_BASE + R_ECX) = R_EDX
```

**Comment by pzread** [comment]
It seems that GDT, LDT 64bit register still need a special case when translating VEX IR to BAP IR.
I have sent a PR to fix this problem.
https://github.com/Cr4sh/openreil/pull/11

**Comment by Cr4sh** [comment]
What exactly are you trying to fix, is there any test case to trigger the bug? `0x65, 0x8B, 0x00` is works without patch.

**Comment by Cr4sh** [comment]
```
reader = ReaderRaw(ARCH_X86, '\x65\x8B\x00')
tr = CodeStorageTranslator(reader)

print tr.get_insn(0)
```
```
;
; asm: mov eax, dword ptr gs:[eax]
; data (3): 65 8b 00
;
00001337.00     STR          R_GS:16,                 ,          V_00:16
00001337.01      OR          V_00:16,             0:16,          V_01:32
00001337.02     STR         R_EAX:32,                 ,          V_02:32
00001337.03     ADD     R_GS_BASE:32,          V_02:32,          V_03:32
00001337.04     STR          V_03:64,                 ,          V_04:32
00001337.05     LDM          V_04:32,                 ,          V_05:32
00001337.06     STR          V_05:32,                 ,         R_EAX:32
```

**Comment by pzread** [comment]
The following code for `65 8b 00` doesn't work on my computer without the patch.
https://gist.github.com/anonymous/5609d3bf491107e96a9d

before patch
```
Error while processing instruction at address 0x0
Exception occurs: panic(): Unhandled register type (I64)
```

after patch
```
00000000.00     STR          R_GS:16,                 ,          V_00:16  
00000000.01      OR          V_00:16,             0:16,          V_01:32  
00000000.02     STR         R_EAX:32,                 ,          V_02:32  
00000000.03     ADD     R_GS_BASE:32,          V_02:32,          V_03:32  
00000000.04     STR          V_03:64,                 ,          V_04:32  
00000000.05     LDM          V_04:32,                 ,          V_05:32  
00000000.06     STR          V_05:32,                 ,         R_EAX:32
```

**Comment by Cr4sh** [comment]
Ok, I figured that this error is reproducible only on 64-bit systems. I'll investigate this case, it looks like libVEX bug because in theory it should operate in the same way on any architecture.

**Comment by Cr4sh** [comment]
Here is VEX code for `mov edi, dword ptr gs:[0x30]` i386 instruction that was generated on 32-bit system:
```
IRSB {
   t0:I32   t1:I32   t2:I32   t3:I64   t4:I32   t5:I32   t6:I16   t7:I64
   t8:I1   t9:I32   t10:I32   t11:I32   t12:I32

   ------ IMark(0x1337, 7, 0) ------
   t6 = GET:I16(296)
   t5 = 16Uto32(t6)
   t1 = GET:I32(300)
   t2 = GET:I32(304)
   t7 = x86g_use_seg_selector{0xa6b10b10}(t1,t2,t5,0x30:I32):I64
   t9 = 64HIto32(t7)
   t8 = CmpNE32(t9,0x0:I32)
   if (t8) { PUT(68) = 0x1337:I32; exit-MapFail }
   t10 = 64to32(t7)
   t11 = LDle:I32(t10)
   PUT(36) = t11
   PUT(68) = 0x133E:I32; exit-Boring
}
```
... and here is VEX code for the same i386 instruction that was generated on 64-bit system:
```
IRSB {
   t0:I32   t1:I64   t2:I64   t3:I64   t4:I32   t5:I32   t6:I16   t7:I64
   t8:I1   t9:I32   t10:I32   t11:I32   t12:I32

   ------ IMark(0x0, 7, 0) ------
   t6 = GET:I16(294)
   t5 = 16Uto32(t6)
   t1 = GET:I64(304)
   t2 = GET:I64(312)
   t7 = x86g_use_seg_selector{0x7fabefaa6cb0}(t1,t2,t5,0x30:I32):I64
   t9 = 64HIto32(t7)
   t8 = CmpNE32(t9,0x0:I32)
   if (t8) { PUT(68) = 0x0:I32; exit-MapFail }
   t10 = 64to32(t7)
   t11 = LDle:I32(t10)
   PUT(36) = t11
   PUT(68) = 0x7:I32; exit-Boring
}
```
As you can see, `t1` and `t2` variables that used to store GDT and LDT addresses has different size on 32 and 64 bit host architectures and it's wrong because guest architecture is still the same.

**Comment by Cr4sh** [comment]
Ok, I figured that technically it's not a bug, libVEX was designed as part of DBI engine, so, from DBI point of view such behaviour has a perfect sense: on 64-bit operating systems GDT and LDT addresses has 64-bit of actual length even for 32-bit user mode code (see [handleSegOverride](https://github.com/Cr4sh/openreil/blob/master/VEX/priv/guest_x86_toIR.c#L142) function of libVEX). 
Check latest commit, [I added your fix for libasmir code](https://github.com/Cr4sh/openreil/blob/master/libasmir/src/irtoir-i386.cpp#L812) with amd64 preprocessor condition as workaround. It should be safe, because LDT/GDT access code that we talking about will be eliminated in any case during trash cleanup after the translation of x86g_use_seg_selector VEX call.
